### ISSUE TITLE
Patterns for Components

### ISSUE THREAD

**Issue by AutoSponge** [WebReflection/viperHTML#5]
I tried to make some simple components with viperhtml where the parent just references `children`. https://glitch.com/edit/#!/comfortable-nail

My first attempt let the component provide its own `wire` if one wasn't provided. This worked well and was really convenient for `map`ing but I noticed re-renders got increasingly more expensive since the `wire` wasn't reused.

So, my second attempt used a WeakMap to hold references to the wires and child models rendered.

In my third attempt, https://glitch.com/edit/#!/hickory-pail, I switched to a partially applied component signature that curried the `wire`. Then rendering children is an exercise of mapping over the child renderers passing the correct model. This was also a convenient pattern and efficient like the second one.

Additional perf gains could be made by having each component check to see if it needs to re-render. If I did that, I would probably want to reuse model objects as well.

For a more complicated child renderer, it would be more efficient to reuse the child renderer and model. So, went back to a `WeakMap` approach to make sure I reorder the child renderers, not just the model. This would be equivalent to `keyed` elements in other frameworks. I also think it's probably more efficient to make the `children` a lazy getter instead of the "simple" approach of updating the `children` array every time a change happens. I think for a faster render cycle, this would make a difference and work better with `raf`.

Any ideas, observations, or improvements?

**Comment by WebReflection** [comment]
I'm sorry I've read this for the first time just now.

Here the catch: latest `hyperHTML` and `viperHTML` supports weakly linked wires by IDs.

```js
const render = viperHTML.wire(model, ':main');
render`
<ul>${
  model.children.map(child => viperHTML.wire(child, ':li')`
  <li> ${child.name} </li>`)
}</ul>`;
```

That will recycle the same node as long as model and children are the same object (by reference, not by their JSON representation).

This pattern was missing the 30th of April so please double check it would be your solution (it looks like it is to me).

**Comment by AutoSponge** [comment]
I'm not sure how the `':id'` helps. Binding the wire in the render seems to do everything I need.

I updated the example. I also added the`announce()` method on model that forces a render to paragraph components without invoking the render. I fixed the `children` method--only needs to fetch the wire, it doesn't need to pass the model again.

https://glitch.com/edit/#!/cosmic-secure

I like this, I like it a lot.

**Comment by WebReflection** [comment]
if you wire the same model in two different template literals the cache is invalidated every single time and it needs to be parsed again per each call.

TL;DR if you are passing different models to paragraph and article you are fine **but** if you are passing the same model you're better off with

```js
// paragraph.js
const viperHTML = require("viperhtml");
module.exports = (model = {}, render = viperHTML.wire(model, ':p')) => () => render`
<li>${model.title}</li>`;

// article.js
const viperHTML = require("viperhtml");
module.exports = (model = {}, render = viperHTML.wire(model, ':article')) => () => render`
<article>
  <h3> ${model.title} </h3>
  List of ${model.children.length} paragraphs:
  <ul>${model.children}</ul>
</article>`;
```

it's a little, yet very important, detail. Consider that if you don't specify an is, a default one will be created for you anyway so ... if I were you I'd just use IDs too :wink: 

Can we close this now?
